core/src/main/java/fr/sncf/osrd/envelope_sim_infra/MRSP.java

```diff
@@ -1,70 +1,57 @@
-    }
-
-    /** Computes the most restricted speed profile from a list of track ranges */
-    public static Envelope from(
-            List<TrackRangeView> ranges,
-            RollingStock rollingStock,
-            boolean addRollingStockLength,
-            String tag) {
+    /**
+     * Computes the MSRP for a rolling stock on a given path.
+     * @param path corresponding path.
+     * @param rollingStock corresponding rolling stock.
+     * @param addRollingStockLength whether the rolling stock length should be taken into account in the computation.
+     * @param trainTag corresponding train.
+     * @return the corresponding MRSP as an Envelope.
+     */
+    public static Envelope computeMRSP(Path path, RollingStock rollingStock, boolean addRollingStockLength,
+                                       String trainTag) {
         var builder = new MRSPEnvelopeBuilder();
-        var pathLength = 0.;
-        for (var r : ranges)
-            pathLength += r.getLength();
+        var pathLength = toMeters(path.getLength());
 
-        // add a limit for the maximum speed the hardware is rated for
+        // Add a limit for the maximum speed the hardware is capable of
         builder.addPart(EnvelopePart.generateTimes(
                 List.of(EnvelopeProfile.CONSTANT_SPEED, MRSPEnvelopeBuilder.LimitKind.TRAIN_LIMIT),
                 new double[] { 0, pathLength },
                 new double[] { rollingStock.maxSpeed, rollingStock.maxSpeed }
         ));
 
-        var offset = 0.;
-        for (var range : ranges) {
-            if (range.getLength() == 0)
-                continue;
-            var subMap = range.getSpeedSections().subRangeMap(Range.closed(0., range.getLength()));
-            for (var speedRange : subMap.asMapOfRanges().entrySet()) {
-                // compute where this limit is active from and to
-                var interval = speedRange.getKey();
-                var begin = offset + interval.lowerEndpoint();
-                var end = offset + interval.upperEndpoint();
-                if (addRollingStockLength) {
-                    end += rollingStock.length;
-                    end = Math.min(pathLength, end);
-                }
-                var speed = speedRange.getValue().getSpeedLimit(tag);
-                if (Double.isInfinite(speed) || speed == 0)
-                    continue;
+        var offset = addRollingStockLength ? rollingStock.length : 0.;
+        var speedRanges = path.getSpeedRanges(trainTag);
+        for (var speedRange: speedRanges) {
+            // Compute where this limit is active from and to
+            var start = toMeters(speedRange.getLower());
+            var end = Math.min(pathLength, offset + toMeters(speedRange.getUpper()));
+            var speed = toMetersPerSecond(rangeMapEntryToSpeed(speedRange));
+            if (!isInfinite(speed) && speed != 0)
```

Comment:
In this setting, the speed can't be infinite (can't be represented as long). We *shouldn't* have infinite values here.